File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/model/factory/DefaultTestStepFactory.java
package org.skellig.teststep.processing.model.factory;

import org.skellig.teststep.processing.converter.TestDataConverter;
import org.skellig.teststep.processing.converter.TestStepValueConverter;
import org.skellig.teststep.processing.model.TestStep;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;
import java.util.Properties;

public class DefaultTestStepFactory extends BaseTestStepFactory {

    private Collection<TestStepFactory> factories;

    public DefaultTestStepFactory(Properties keywordsProperties,
                                  TestStepValueConverter testStepValueConverter,
                                  Collection<TestStepFactory> factories,
                                  TestDataConverter testDataConverter) {
        super(keywordsProperties, testStepValueConverter, testDataConverter);
        this.factories = factories;
    }

    @Override
    public TestStep create(String testStepName, Map<String, Object> rawTestStep, Map<String, String> parameters) {
        return factories.stream()
                .filter(factory -> factory.isConstructableFrom(rawTestStep))
                .findFirst()
                .map(factory -> factory.create(testStepName, rawTestStep, parameters))
                .orElse(super.create(testStepName, rawTestStep, parameters));
    }

    @Override
    protected TestStep.Builder createTestStepBuilder(Map<String, Object> rawTestStep, Map<String, Object> parameters) {
        return new TestStep.Builder();
    }

    @Override
    public boolean isConstructableFrom(Map<String, Object> rawTestStep) {
        return true;
    }

    public static class Builder {

        private Collection<TestStepFactory> testStepFactories;
        private Properties keywordsProperties;
        private TestStepValueConverter testStepValueConverter;
        private TestDataConverter testDataConverter;

        public Builder() {
            testStepFactories = new ArrayList<>();
        }

        public Builder withTestStepFactory(TestStepFactory factory) {
            this.testStepFactories.add(factory);
            return this;
        }

        public Builder withKeywordsProperties(Properties keywordsProperties) {
            this.keywordsProperties = keywordsProperties;
            return this;
        }

        public Builder withTestStepValueConverter(TestStepValueConverter testStepValueConverter) {
            this.testStepValueConverter = testStepValueConverter;
            return this;
        }

        public Builder withTestDataConverter(TestDataConverter testDataConverter) {
            this.testDataConverter = testDataConverter;
            return this;
        }

        public TestStepFactory build() {
            return new DefaultTestStepFactory(keywordsProperties, testStepValueConverter,
                    testStepFactories, testDataConverter);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/model/factory/DefaultTestStepFactory.kt
package org.skellig.teststep.processing.model.factory

import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.value.ValueExpressionContextFactory
import org.skellig.teststep.reader.value.expression.ValueExpression


/**
 * This class is responsible for creating instances of the [DefaultTestStep] class.
 * It extends the [BaseDefaultTestStepFactory] class and implements the [TestStepFactory] interface
 * with the [DefaultTestStep] type parameter.
 *
 * @param testStepRegistry The test step registry used to store and retrieve test steps.
 * @param valueExpressionContextFactory The factory used to create instances of `ValueExpressionContext`.
 */
internal class DefaultTestStepFactory(
    testStepRegistry: TestStepRegistry,
    valueExpressionContextFactory: ValueExpressionContextFactory
) : BaseDefaultTestStepFactory<DefaultTestStep>(testStepRegistry, valueExpressionContextFactory) {

    override fun createTestStepBuilder(rawTestStep: Map<ValueExpression, ValueExpression?>, parameters: Map<String, Any?>): DefaultTestStep.Builder<DefaultTestStep> {
        return DefaultTestStep.DefaultTestStepBuilder()
    }

    override fun isConstructableFrom(rawTestStep: Map<ValueExpression, ValueExpression?>): Boolean {
        return true
    }

    class Builder {
        private var valueExpressionContextFactory: ValueExpressionContextFactory? = null
        private var testStepRegistry: TestStepRegistry? = null

        fun withValueExpressionContextFactory(valueExpressionContextFactory: ValueExpressionContextFactory?) =
            apply { this.valueExpressionContextFactory = valueExpressionContextFactory }

        fun withTestStepRegistry(testStepRegistry: TestStepRegistry?) =
            apply { this.testStepRegistry = testStepRegistry }

        fun build(): TestStepFactory<DefaultTestStep> {
            return DefaultTestStepFactory(
                testStepRegistry ?: error("TestStepRegistry is mandatory for DefaultTestStepFactory"),
                valueExpressionContextFactory ?: error("ValueExpressionContextFactory is mandatory for DefaultTestStepFactory"),
            )
        }
    }
}